lib/c.c

```diff
@@ -436,80 +440,94 @@ int fputc(int c, FILE *stream)
     return 0;
 }
 
-/* Reference:
- *   https://danluu.com/malloc-tutorial/
- * FIXME: adopt lite_malloc from musl-libc
- *   https://git.musl-libc.org/cgit/musl/tree/src/malloc/lite_malloc.c
+/*  FIXME: Best-fit Algorithm
+ *   https://github.com/picolibc/picolibc
  */
 
-typedef struct block_meta {
+#define PROT_READ 0x01
+#define PROT_WRITE 0x02
+#define MAP_PRIVATE 0x02
+#define MAP_ANONYMOUS 0x20
+#define PAGESIZE 4096
+
+typedef struct _chunk {
+    struct _chunk *next;
+    struct _chunk *prev;
     int size;
-    struct block_meta *next;
-    int free;
-} block_meta_t;
+    void *ptr;
+} chunk;
 
-block_meta_t *__malloc_global_base;
-block_meta_t *__malloc_global_last;
+chunk *head;
 
-block_meta_t *__malloc_request_space(int size)
+void align_up(int size)
 {
-    char *brk;
-    block_meta_t *block;
-    brk = __syscall(__syscall_brk, 0);
-    block = brk;
-
-    char *request =
-        __syscall(__syscall_brk, block + size + sizeof(block_meta_t));
-    if (request == -1)
-        return NULL;
-
-    if (__malloc_global_last)
-        __malloc_global_last->next = block;
+    int mask = PAGESIZE - 1;
 
-    block->size = size;
-    block->next = NULL;
-    block->free = 0;
-    return block;
+    return ((size - 1) | mask) + 1;
 }
 
 void *malloc(int size)
 {
-    block_meta_t *block;
-    if (size == 0)
+    int flags = MAP_PRIVATE | MAP_ANONYMOUS;
```

Comment:
At the moment, there is NO optimizer, which means that additional computation at runtime is required for bitwise-OR and assignment. Thus, you shall pre-calculate these values (with proper comments) and specify them in mmap directly.